Show addressee typeahead when starting a new DM with no recipient.

pratik-pc opened the pull request:
This PR:
- adds offset for tippy placement to typeahead and moves DM typeahead `6px` above the input.
- Make shift+tab work as keyboard navigation when typeahead is displayed
- Whenever `#private_message_recipient` is brought to `focus` and is empty, we show the DM recipient typeahead

Fixes: zulip#31803.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![Screenshot from 2024-11-26 00-40-19](https://github.com/user-attachments/assets/477a1220-a2eb-47a9-b115-e236e1845a85)

[Screencast from 25-11-24 11:38:53 PM IST.webm](https://github.com/user-attachments/assets/ec2fbd43-ba2f-47e7-ae5d-7dc3f22dafc4)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Seems worth trying to me! Works as expected in my manual testing.

@evykassirer are you up for reviewing this one?

evykassirer commented:
Here's what can hopefully be a simpler version of your second commit:

```diff
diff --git a/web/src/compose_actions.ts b/web/src/compose_actions.ts
index dacca07fa2..8a8d9ae4b1 100644
--- a/web/src/compose_actions.ts
+++ b/web/src/compose_actions.ts
@@ -325,6 +325,11 @@ export let start = (raw_opts: ComposeActionsStartOpts): void => {
         clear_box();
     }
 
+    // Set the recipients with a space after each comma, so it looks nice.
+    compose_state.private_message_recipient(
+        opts.private_message_recipient.replaceAll(/,\s*/g, ", "),
+    );
+
     if (opts.message_type === "private") {
         compose_state.set_compose_recipient_id(compose_state.DIRECT_MESSAGE_ID);
         compose_recipient.on_compose_select_recipient_update();
@@ -338,11 +343,6 @@ export let start = (raw_opts: ComposeActionsStartOpts): void => {
     }
     compose_state.topic(opts.topic);
 
-    // Set the recipients with a space after each comma, so it looks nice.
-    compose_state.private_message_recipient(
-        opts.private_message_recipient.replaceAll(/,\s*/g, ", "),
-    );
-
     // If we're not explicitly opening a different draft, restore the last
     // saved draft (if it exists).
     let restoring_last_draft = false;
diff --git a/web/src/compose_ui.ts b/web/src/compose_ui.ts
index 034a17a50f..8400414a8f 100644
--- a/web/src/compose_ui.ts
+++ b/web/src/compose_ui.ts
@@ -180,6 +180,9 @@ export function set_focus(opts: ComposeTriggeredOptions): void {
     if (window.getSelection()!.toString() === "" || opts.trigger !== "message click") {
         const focus_area = get_focus_area(opts);
         $(focus_area).trigger("focus");
+        if (focus_area === "#private_message_recipient") {
+            $(focus_area).trigger("click");
+        }
     }
 }

```

It seemed like the thing you were trying to avoid was the typeahead flashing when the composebox opens with a DM participant already in the recipient box. (If not, let me know and I can help find cleaner ways to address those issues.)

Instead of adding extra if statements, which are a bit hard to follow why they're there, this diff I came up with makes sure that we always click the DM recipient box if it's the area of focus. And to avoid the flash, it sets the recipient in the box before getting to the code that checks for the focus area, so that we skip trying to focus the recipient box at all if there's already a recipient.

Let me know what you think!

pratik-pc commented:
> this diff I came up with makes sure that we always click the DM recipient box if it's the area of focus. And to avoid the flash, it sets the recipient in the box before getting to the code that checks for the focus area, so that we skip trying to focus the recipient box at all if there's already a recipient.

Cool, this is better

evykassirer commented on web/src/composebox_typeahead.ts:
```diff
@@ -596,6 +596,9 @@ export function filter_and_sort_mentions(
 export function get_pm_people(query: string): (UserGroupPillData | UserPillData)[] {
+    if (query === "" && $("#private_message_recipient").siblings().length > 0) {
+        return [];
+    }
```
I think you should be able to remove this now

pratik-pc replied:
I wrote this check to make sure typeahead appears only when there is no recipient with empty string. I though this was the intended design.(?) Subsequent empty strings would open typeahead on click

evykassirer replied:
Can you explain what would happen if this code wasn't here? Like what could the user do to end up in an undesirable state?

pratik-pc replied:
If a recipient user pill is already present, clicking inside the recipient box (possibly to add another user) will open up typeahead since we have `helponemptrystring` set to true. I though it was only for empty recipient box(?)
>we should try immediately opening the addressee typeahead when there is no recipient selected

evykassirer replied:
I think that might be desirable behavior actually. Can you ask on CZO to see if people prefer the typeahead opening or not?

alya commented:
Looks good to me in manual testing!

alya commented:
@timabbott This PR was reviewed by @evykassirer , and I tested it as well.

timabbott commented:
I noticed #32912 while reviewing this, but I've confirmed that bug was present in `main`, so it's not an issue for this.

The one thing I worry about with this PR is if the typeahead should sit a few pixels higher, so it isn't overlapping the border of the text input row. So maybe we should look at pushing that up? I'ved [asked in #design](https://chat.zulip.org/#narrow/channel/101-design/topic/pm.20compose.20should.20open.20typeahead/near/2027782).

![image](https://github.com/user-attachments/assets/f7dce9c8-4910-48c8-85c7-107fdc4b89f4)

timabbott commented:
Update the deployment of this on chat.zulip.org.

pratik-pc commented:
Adds shift tabbing handler for `private_message_recipient` whenever typeahead is open.

Also fixed auto select for typeahead item whenever focus is brought to `private_message_recipient` from dropdown on `keydown`.

timabbott commented:
@pratik-pc is this ready to test again aside from the merge conflicts? If so, please add the "integration review" label after rebasing.

pratik-pc commented:
@zulipbot add "integration review"

timabbott commented on web/src/compose_recipient.ts:
```diff
@@ -340,6 +341,19 @@ export function initialize(): void {
+                    composebox_typeahead.toggle_ignore_typeahead_keyup(true);
+                }
+            });
+        },
```
Can you add a block comment for this block explaining exactly what it does and why?

pratik-pc commented:
> I was able to reproduce the first user getting randomly selected when doing "Shift+Tab" to move from the compose textarea into the DM recipient input. I think that likely suggests the `ignore_keyup` logic isn't robust.
> 
> It felt pretty hacky to begin with; I wonder if we just need a different strategy.

@timabbott Now, we just track whether the user meant to select a typeahead item or if the focus simply shifted to the typeahead between the keydown and keyup events.

pratik-pc commented:
Rebased onto main

amanagr committed:
typeahead: Add showOnFocus option.

This adds a new option to the Typeahead class, following the same
pattern as the existing showOnClick option. When showOnFocus is
true, the typeahead will call lookup() when the input element
gains focus, allowing it to display suggestions immediately.

Co-authored-by: Anirudh Singh Rajora <anirudh.s.me@gmail.com>

pratik-pc committed:
composebox_typeahead: Offset private message recipient typeahead.

We add a `positionOffset` option to bootstrap typeahead to define
offset value to tippy instances that shifts the tippy popover by
[x, y] pixels relative to its default placement.

Fixes: https://chat.zulip.org/#narrow/channel/101-design/topic/pm.20compose.20should.20open.20typeahead/near/2027782

pratik-pc committed:
bootstrap_typeahead: Drive query updates from the input event.

Previously, the keyup handler was responsible for both navigation
(Tab/Enter/Escape selection) and refreshing the query after the user
typed. Using keyup for the latter caused trouble when keyboard focus
moved between elements mid-keystroke: the keydown fires on the old
element and the keyup on the newly focused one. For a typeahead input
that opens on focus, the stray Tab or Shift keyup landed in the
default branch and interfered with the just-opened typeahead, which
is why we previously had a special case for Shift on the topic field.

We now listen for the `input` event, which fires only when the input's
value actually changes. Focus-transition keyups no longer reach the
query-update path at all, so the Shift-on-topic workaround is no
longer needed, and the long-standing NOTE about Command+Delete on Mac
(which doesn't emit a keyup for Delete) also goes away -- `input`
fires directly on the deletion. Backspace is detected via
`InputEvent.inputType === "deleteContentBackward"` to preserve the
existing `hideOnEmptyAfterBackspace` behavior.

`mouse_moved_since_typeahead` is reset in keydown now, so it still
runs once per key interaction regardless of whether that keystroke
produces a value change.

The `openInputFieldOnKeyUp` option is renamed to
`openInputFieldOnInput` to match the new event.

Co-Authored-By: Evy Kassirer <evy@zulip.com>

pratik-pc committed:
bootstrap_typeahead: Don't select from typeahead on shift+tab.

The default browser behaviour for shift+tab is backward focus navigation.
The bootstrap typeahead currently selects typeahead item on shift+tab
overriding the navigation behaviour.

We return early before calling `select()` to preserve this default browser
navigation behaviour.

pratik-pc committed:
composebox_typeahead: Open DM recipient typeahead when field is empty.

When composing a new DM with no recipients, automatically open the
typeahead so users can see suggested recipients. We enable `showOnFocus`
and `helpOnEmptyStrings` for the private message recipient typeahead so
it is shown when the field receives focus and the query is empty.

Also adds a check in `get_pm_people()` to return early for an empty
query when a user pill is already present, preventing the typeahead
from being shown in that case.

Fixes: zulip#31803.

evykassirer commented on web/src/bootstrap_typeahead.ts:
```diff
@@ -255,6 +255,7 @@ export class Typeahead<ItemType extends string | object> {
     closeInputFieldOnHide: (() => void) | undefined;
     helpOnEmptyStrings: boolean;
     tabIsEnter: boolean;
+    shiftTabIsEnter: boolean;
```
Instead of making this option, could we just change this for all typeaheads? Seems worth starting a frontend or design thread for. I'd prefer to keep things simpler instead of making more options in the typeahead code, if possible.

pratik-pc replied:
Yeah, I think it makes sense to change this for all typeaheads since user probably wants to navigate if they are doing `Shift+tab`

evykassirer replied:
Can you start a design to thread to double check if people are okay with this behavior change? This could also be a prep PR to make the change.

evykassirer replied:
Nevermind, trying it out locally it's obvious to me that this is a good behavior change :)

evykassirer commented on web/src/bootstrap_typeahead.ts:
```diff
@@ -245,6 +245,8 @@ export class Typeahead<ItemType extends string | object> {
     mouse_moved_since_typeahead = false;
     shown = false;
+    // Tracks which keys had a keydown event fire on this element.
+    keydown_set = new Set<string>();
```
This feels hacky to me. I'd like to avoid creating new typeahead options where we can instead just handle events elegantly.

I asked Claude for an alternative, and it wrote https://github.com/evykassirer/zulip/pull/36 -- can you check that out and see what you think? I haven't tested it or looked thoroughly into it, but it looks promising

pratik-pc replied:
Yeah, if we're okay migrating from `keydown` to `input` (though they're essentially the same behavior-wise), then this is better.

pratik-pc replied:
Actually, this doesn't solve the issue. "Focus-transition keyups no longer reach the query-update path at all."(commit message excerpt) We aren't particularly interested in deferring the lookup here to input, since we're implementing lookup on focus anyway. Instead, we want to avoid executing select() during such focus transitions. I missed this during testing because we changed `Shift+Tab` to no longer trigger this path, but on closer inspection, normal Tab navigation still has this issue.

pratik-pc replied:
```diff
diff --git a/web/src/bootstrap_typeahead.ts b/web/src/bootstrap_typeahead.ts
index fcfaa874e2..5e1c213ede 100644
--- a/web/src/bootstrap_typeahead.ts
+++ b/web/src/bootstrap_typeahead.ts
@@ -238,6 +238,12 @@ export class Typeahead<ItemType extends string | object> {
     query = "";
     mouse_moved_since_typeahead = false;
     shown = false;
+    // Whether a keydown has been seen on this element since it gained
+    // focus. Selection keyups (e.g. Tab) only act when this is true, so
+    // a stray keyup whose keydown fired on a different element -- as
+    // happens when focus moves into this input via keyboard navigation
+    // -- doesn't trigger a selection.
+    keydown_seen_since_focus = false;
     // To trigger updater when Esc is pressed only during the stream topic typeahead in composebox.
     select_on_escape_condition: () => boolean;
     // Used to clear tooltip instances attached to typeahead container.
@@ -772,6 +778,8 @@ export class Typeahead<ItemType extends string | object> {
     }
 
     keydown(e: JQuery.KeyDownEvent): void {
+        this.keydown_seen_since_focus = true;
+
         // Any key interaction should give keyboard priority over the
         // mouse for selecting the currently highlighted typeahead item.
         this.mouse_moved_since_typeahead = false;
@@ -807,8 +815,18 @@ export class Typeahead<ItemType extends string | object> {
         // input via keyboard navigation).
         switch (e.key) {
             case "Tab":
-                // If the typeahead is not shown or tabIsEnter option is not set, do nothing and return
-                if (!this.tabIsEnter || !this.shown || e.shiftKey) {
+                // Do nothing when the typeahead isn't shown, tabIsEnter
+                // is off, or this is a shift+tab. Also bail when we
+                // haven't seen a keydown since focus: that means this
+                // keyup belongs to a Tab whose keydown fired on another
+                // element while focus was moving into this input, and
+                // it must not select a typeahead item.
+                if (
+                    !this.tabIsEnter ||
+                    !this.shown ||
+                    e.shiftKey ||
+                    !this.keydown_seen_since_focus
+                ) {
                     return;
                 }
 
@@ -904,6 +922,7 @@ export class Typeahead<ItemType extends string | object> {
     }
 
     element_focus(): void {
+        this.keydown_seen_since_focus = false;
         if (!this.showOnFocus) {
             return;
         }
```


maybe something like this or moving tab select out of `keyup`

evykassirer replied:
Thanks for looking into it! Yup, Claude does sometimes (often lol) make mistakes. I think your proposed diff or moving tab select out of keyup could both work, and I'd prefer moving tab select out of `keyup` since that feels like it would be simpler to reason about in the end and have less state being kept track of in the typeahead object. Thoughts?

evykassirer commented on web/src/compose_recipient.ts:
```diff
@@ -205,7 +205,7 @@ export function update_on_recipient_change(): void {
 }
 
-function switch_message_type(message_type: MessageType): void {
+function switch_message_type(message_type: MessageType, defer_focus = false): void {
```
It would be nice if we could avoid having the `defer_focus` option, since that adds extra complexity. I asked Claude "is there a way to achieve the goal without defer_focus?", and it said:

> Yes — there's a cleaner alternative: move the `set_focus` call out of `switch_message_type` entirely.                              
>                                                                                                                                  
> Looking at all the callers of `on_compose_select_recipient_update`:                                                              
> 
> - compose_actions.start (×3) — focus is handled separately by `show_compose_box` (sync or via setTimeout when deferring).        
> - `item_click_callback` (user clicks dropdown) — genuinely needs focus to move, since the field that was focused may now be hidden.                                                                                                                        
> - server_events_dispatch, stream_events (×2), stream_settings_ui, message_events, possibly_update_stream_name_in_compose —       triggered by server events/settings changes; stealing focus from whatever the user is doing would be a bug.                    
>                                                                                                                                 
>  So of seven callers, only one actually wants switch_message_type to steal focus. The current design opts-out via defer_focus;  
>  the inverse is cleaner — switch_message_type doesn't set focus, and the one caller that needs it (item_click_callback) calls   
>  `compose_ui.set_focus` itself.                              
>                                                                                                                                 
>  That lets you drop `defer_focus` entirely (from both compose_recipient.ts and the three call sites in compose_actions.ts), fixes  a latent bug where server-event-triggered type switches could steal focus, and the reordering of `set_private_message_recipient_ids` — which is the actual fix the commit title describes — stands on its own as a single clean  commit.   

That would look like:

```diff
diff --git a/web/src/compose_actions.ts b/web/src/compose_actions.ts
index 074f40ea79..4c42b6b926 100644
--- a/web/src/compose_actions.ts
+++ b/web/src/compose_actions.ts
@@ -387,6 +387,8 @@ export let start = (raw_opts: ComposeActionsStartOpts): void => {
         clear_box();
     }
 
+    compose_state.set_private_message_recipient_ids(opts.private_message_recipient_ids);
+
     if (opts.message_type === "private") {
         compose_state.set_compose_recipient_id(compose_state.DIRECT_MESSAGE_ID);
         compose_recipient.on_compose_select_recipient_update();
@@ -426,8 +428,6 @@ export let start = (raw_opts: ComposeActionsStartOpts): void => {
     }
     compose_recipient.update_topic_displayed_text(opts.topic);
 
-    compose_state.set_private_message_recipient_ids(opts.private_message_recipient_ids);
-
     // If we're not explicitly opening a different draft, restore the last
     // saved draft (if it exists).
     let restoring_last_draft = false;
diff --git a/web/src/compose_recipient.ts b/web/src/compose_recipient.ts
index 6e05c6b1a5..dc08dcd5ea 100644
--- a/web/src/compose_recipient.ts
+++ b/web/src/compose_recipient.ts
@@ -205,21 +205,40 @@ export function update_on_recipient_change(): void {
     compose_validate.clear_topic_moved_info();
 }
 
+function current_compose_triggered_opts(
+    message_type: MessageType,
+    trigger: string,
+): ComposeTriggeredOptions {
+    if (message_type === "private") {
+        return {
+            message_type: "private",
+            trigger,
+            private_message_recipient_ids: compose_state.private_message_recipient_ids(),
+        };
+    }
+    return {
+        message_type: "stream",
+        trigger,
+        stream_id: compose_state.stream_id()!,
+        topic: compose_state.topic(),
+    };
+}
+
 function switch_message_type(message_type: MessageType): void {
     $("#compose-content .alert").hide();
 
     compose_state.set_message_type(message_type);
 
-    const opts = {
-        message_type,
-        trigger: "switch_message_type",
-        stream_id: compose_state.stream_id()!,
-        topic: compose_state.topic(),
-        private_message_recipient_ids: compose_state.private_message_recipient_ids(),
-    };
-    update_compose_for_message_type(opts);
+    update_compose_for_message_type(
+        current_compose_triggered_opts(message_type, "switch_message_type"),
+    );
     update_compose_area_placeholder_text();
-    compose_ui.set_focus(opts);
+    // Focus is intentionally not moved here. Callers that want to
+    // move focus after a message-type switch (e.g., the recipient
+    // dropdown click handler) are responsible for doing so; callers
+    // that fire on server events or during compose-box opening should
+    // not disturb focus, which is either already where it belongs or
+    // will be set by `show_compose_box`.
 }
 
 function update_recipient_label(stream_id?: number): void {
@@ -312,7 +331,17 @@ function item_click_callback(event: JQuery.ClickEvent, dropdown: tippy.Instance)
     compose_state.set_recipient_edited_manually(true);
     // Enable or disable topic input based on `topics_policy`.
     update_topic_displayed_text(compose_state.topic());
+    const prev_message_type = compose_state.get_message_type();
     on_compose_select_recipient_update();
+    const message_type = compose_state.get_message_type();
+    if (message_type !== undefined && prev_message_type !== message_type) {
+        // A message-type switch hides the field that was previously
+        // focused (topic for stream, DM input for private), so move
+        // focus into the first empty input of the new compose state.
+        compose_ui.set_focus(
+            current_compose_triggered_opts(message_type, "switch_message_type"),
+        );
+    }
     compose_select_recipient_dropdown_widget.item_clicked = true;
     dropdown.hide();
     event.preventDefault();
```

What do you think?

pratik-pc replied:
This looks great.

evykassirer commented on web/src/composebox_typeahead.ts:
```diff
@@ -1830,5 +1830,35 @@ export function initialize({
     });
 
+    // Track whether focus arrived via a click so the focus handler can skip it.
+    let recipient_focused_via_click = false;
```
`recipient_focused_via_click` feels a bit hacky. What about this instead?

```diff
diff --git a/web/src/composebox_typeahead.ts b/web/src/composebox_typeahead.ts
index 771408cb5c..f5b2d6c05c 100644
--- a/web/src/composebox_typeahead.ts
+++ b/web/src/composebox_typeahead.ts
@@ -1826,6 +1826,23 @@ export function initialize({
             set_recipient_from_typeahead(item);
         },
         stopAdvance: true, // Do not advance to the next field on a Tab or Enter
+        // Show suggestions even when the field is empty, so that focusing
+        // an empty recipient input (e.g., when starting a new DM) surfaces
+        // candidate recipients without the user having to type first.
+        helpOnEmptyStrings: true,
+    });
+
+    // Open the DM recipient typeahead when the field is focused and empty,
+    // so a user starting a new DM immediately sees suggested recipients.
+    // With `helpOnEmptyStrings: true` above, a plain `lookup(false)` on an
+    // empty field shows suggestions instead of hiding, so the
+    // focus-then-click sequence on the input cooperates naturally --
+    // bootstrap_typeahead's own `element_click` lookup re-shows the
+    // suggestions rather than closing them.
+    $("#compose-direct-recipient").on("focus", "#private_message_recipient", () => {
+        if (compose_state.private_message_recipient_ids().length === 0) {
+            private_message_recipient_typeahead.lookup(false);
+        }
     });
 
     initialize_compose_typeahead($("textarea#compose-textarea"));
```

pratik-pc replied:
I added an extra condition in `get_pm_people` so that the typeahead is not shown for an empty query when a user pill already exists. I also cherry-picked the showOnFocus commit from #37060, since it seems worthwhile to add a dedicated option now that we'll be using this behavior for multiple typeaheads.

pratik-pc committed:
compose_recipient: Move focus handling out of `switch_message_type`.

`switch_message_type()` was responsible for moving focus after every
message type change. However, only the compose recipient dropdown click
handler actually requires focus to be moved. Other callers, including
compose initialization and server event updates, should not steal
focus from the user.

We refactor the `switch_message_type` logic to leave focus unchanged
and move the focus handling into the dropdown click handler, where the
appropriate input field in the updated compose box is focused.

Also calls `compose_state.set_private_message_recipient_ids()` before
`on_compose_select_recipient_update()` fixing a momentary flash where
`#compose-recipient` receives the `.recently-focus` class.

pratik-pc committed:
search: Remove dead openInputFieldOnInput code.

`openInputFieldOnInput` in the search typeahead was used to open the
search input when the user started typing while the search box was
closed.
In commit 75dab825feac0abc051f59e0e75488b7e8aa0b74, we removed the
ability to type into closed search box, making this code path
unreachable.

pratik-pc committed:
bootstrap_typeahead: Refocus input on click event instead of blur.

We remove the fragile blur handler logic introduced in commit
1b86a8f398cd6e6bc21180e23325c7e49d40fc66 that refocused the input
element when the user clicked on a typeahead header. Instead, we handle
the click event directly, check whether the user selected a typeahead
item, and only refocus the input if they did not. This also avoids a
double focus issue that would otherwise occur once the showOnFocus
option is introduced in a later commit.

pratik-pc commented on web/src/search.ts:
```diff
@@ -269,11 +269,6 @@ export function initialize(opts: {on_narrow_search: OnNarrowSearch}): void {
-            if ($(".navbar-search.expanded").length === 0) {
-                open_search_bar_and_close_narrow_description();
-            }
-        },
```
`openInputFieldOnKeyUp` is dead code after 75dab825feac0abc051f59e0e75488b7e8aa0b74. Should we just remove the option from the typeahead altogether?

pratik-pc commented:
@evykassirer addressed above feedback

evykassirer commented on web/src/bootstrap_typeahead.ts:
```diff
@@ -267,6 +267,9 @@ export class Typeahead<ItemType extends string | object> {
     // Used to determine whether the typeahead should be shown,
     // when the user clicks anywhere on the input element.
     showOnClick: boolean;
+    // Used to determine whether the typeahead should be shown,
```
nit: no comma needed at the end of this line

evykassirer commented on web/src/bootstrap_typeahead.ts:
```diff
@@ -852,33 +854,28 @@ export class Typeahead<ItemType extends string | object> {
+        }
+        const input_event = e.originalEvent;
+        const is_backspace =
+            input_event instanceof InputEvent && input_event.inputType === "deleteContentBackward";
```
I asked Claude about this change from `e.key === "Backspace"` to using `InputEvent` and `inputType`, and it said:

---

This narrows the backspace detection compared to the `keyup` version it
replaces, and I think it's a behavior regression for `hideOnEmptyAfterBackspace`.

The old code checked `e.key === "Backspace"`. On macOS, **Option+Backspace**
(word delete) and **Cmd+Backspace** (line delete) both report
`e.key === "Backspace"`, so the previous behavior hid the typeahead when any
of those emptied the field. `inputType === "deleteContentBackward"` only
matches a single-character backspace, so word/line deletes to empty now fall
through to `lookup(false)` and the typeahead stays open.

Since all the backward-deletion input types end in `"Backward"`
(`deleteContentBackward`, `deleteWordBackward`, `deleteSoftLineBackward`,
`deleteHardLineBackward`), we can restore parity with:

```js
const input_event = e.originalEvent;
// Treat any backward deletion (plain Backspace, plus Option/Cmd+Backspace
// word/line deletes) the way the old `e.key === "Backspace"` check did.
const is_backspace =
    input_event instanceof InputEvent &&
    input_event.inputType.startsWith("delete") &&
    input_event.inputType.endsWith("Backward");
```

I'd keep it to backward deletions only (not forward Delete or cut), to match
the `hideOnEmptyAfterBackspace` name and the old key-based behavior.

---

It feels a bit hacky of a solution, but probably is okay? It will help to make this logic a separate helper function (maybe in `util.ts`?) for checking if an input event is backspace, since that could be used elsewhere too.

evykassirer commented on web/src/compose_recipient.ts:
```diff
@@ -210,21 +210,39 @@ export function update_on_recipient_change(): void {
+    // dropdown click handler) are responsible for doing so; callers
+    // that fire on server events or during compose-box opening should
+    // not disturb focus, which is either already where it belongs or
+    // will be set by `show_compose_box`.
```
The commit message says "Other callers, including compose initialization and server event updates, should not steal focus from the user." -- I asked Claude about this since I was surprised that we'd want to switch recipient and not focus the composebox, and eventually it said:

---

The motivation here is sound, but I think the commit message overstates the
server-event part. `switch_message_type` only runs when the message type
actually flips (`prev_message_type !== curr_message_type` in
`on_compose_select_recipient_update`), and `curr` is `"private"` only when
`selected_recipient_id === "direct"`.

Tracing the non-interactive callers, none of them can trigger a flip:

- `stream_events.ts` and `server_events_dispatch.js` (stream archived/deleted)
  call `set_selected_recipient_id("")` — `""` is still stream-typed, and
  they're guarded by `if (stream_id === compose_state.selected_recipient_id)`,
  so they only fire when you're already composing to that stream. They can't
  reach the DM state, so `prev === curr` and `switch_message_type` is never
  called.
- The stream privacy/settings updates don't change the selected recipient at
  all.

So the type only actually flips on (1) the recipient dropdown click — where we
*do* want focus moved — and (2) compose open via `start()`. That means the real,
observable win of this change is eliminating the **double `set_focus` on compose
open**: both `show_compose_box` and `switch_message_type` were focusing the box,
which is exactly the duplication that `defer_focus` was papering over.

Could we lead the commit message with that, and frame the server-event /
non-interactive callers as defensive correctness ("shouldn't steal focus if a
future caller flips the type from a background context") rather than implying
they steal focus today? As written it reads like it's fixing an active
focus-steal on server events, which the current code can't actually hit.

evykassirer commented on web/src/compose_actions.ts:
```diff
@@ -387,6 +387,8 @@ export let start = (raw_opts: ComposeActionsStartOpts): void => {
         clear_box();
     }
 
+    compose_state.set_private_message_recipient_ids(opts.private_message_recipient_ids);
```
This seems like a separate bug fix that would benefit from its own commit (or PR?) with clear before/after visuals, and an explanation of why this change fixes it and why it doesn't cause further bugs.

evykassirer commented on web/src/composebox_typeahead.ts:
```diff
@@ -1891,6 +1897,10 @@ export function initialize({
             set_recipient_from_typeahead(item);
         },
         stopAdvance: true, // Do not advance to the next field on a Tab or Enter
+        // show suggested recipients as soon as the field is focused with no query,
```
Capitalize "Show" here to make it clearer this is the start of a new sentence.

evykassirer commented on web/src/search.ts:
```diff
@@ -269,11 +269,6 @@ export function initialize(opts: {on_narrow_search: OnNarrowSearch}): void {
             exit_search({keep_search_narrow_open: false});
         },
         tabIsEnter: true,
-        openInputFieldOnInput(): void {
```
This is a great find that there's some dead code, but I think removing it would also be best as a separate PR (to be merged first). That would also let us avoid the rename from `openInputFieldOnKeyUp` altogether. And yeah removing it completely from the bootstrap typeahead file seems fine too.

evykassirer reviewed (commented):
I looked through the code and it seems like review was addressed well and the code is in good shape. But this has turned into quite the involved PR!

The `bootstrap_typeahead` code is tricky to review because it affects several different typeaheads and the code itself is a bit less nice overall because it came from boostrap. Your three `bootstrap_typeahead` commits change things for _all_ typeaheads, which is probably what we want to do to keep the code simpler, but I'm thinking they would be best reviewed separately as prep PRs so we can more confidently confirm they don't cause regressions and that they change behavior in the way we want. Same with the `compose_recipient: Move focus handling out of switch_message_type` commit. And then in each of these PRs, you can add some clear screenshots showing the changes, describe what you manually tested, and ensure there are automated tests where appropriate.
